File: Python/Sorting Algorithm/Merge_Sort/merge_sort_using_array.py
def merge(arr, lower_bound, mid, upper_bound):
    new_arr = []
    i = lower_bound
    j = mid + 1
    while i <= mid and j <= upper_bound:
        if arr[i] < arr[j]:
            new_arr.append(arr[i])
            i += 1
        else:
            new_arr.append(arr[j])
            j += 1
    while i <= mid:
        new_arr.append(arr[i])
        i += 1
    while j <= upper_bound:
        new_arr.append(arr[j])
        j += 1

    j = 0
    i = lower_bound
    while i <= upper_bound:
        arr[i] = new_arr[j]
        j += 1
        i += 1


def merge_sort(arr, lower_bound, upper_bound):
    if lower_bound < upper_bound:
        mid = (lower_bound + upper_bound) // 2
        merge_sort(arr, lower_bound, mid)
        merge_sort(arr, mid + 1, upper_bound)

        merge(arr, lower_bound, mid, upper_bound)


if __name__ == "__main__":
    n = int(input('\nEnter No of Input :'))
    arr = [int(input('Enter array element :')) for i in range(n)]
    print("\nBefore Sort :", arr)
    merge_sort(arr, 0, n-1)
    print("\nAfter Sort :", arr)

File: C++/Sorting Algorithm/Merge_Sort/merge_sort_using_array.cpp
#include<iostream>
using namespace std;
void merge(int arr[], int lower_bound, int mid, int upper_bound){

	int *temp_arr = (int*)malloc((upper_bound + 1) * sizeof(int));
	for ( int i = lower_bound; i <= upper_bound; i ++){
		temp_arr[i] = arr[i];
	}
	int i = lower_bound;
	int j = mid + 1;
	int k = lower_bound;

	while (i <= mid and j <= upper_bound){
		if (temp_arr[i] < temp_arr[j])
			arr[k++] = temp_arr[i++];
		else 
			arr[k++] = temp_arr[j++];	
	}	
	while (i <= mid)
		arr[k++] = temp_arr[i++];
	while (j <= upper_bound)
		arr[k++] = temp_arr[j++];

	free(temp_arr);
}
void merge_sort(int arr[], int lower_bound, int upper_bound){
	if (lower_bound < upper_bound){
		int mid = (lower_bound + upper_bound) / 2;
		merge_sort(arr, lower_bound, mid);
		merge_sort(arr, mid + 1, upper_bound);
		merge(arr, lower_bound, mid, upper_bound);
	}
}

void print(int arr[], int n){
	for (int i = 0; i < n; i++)
		cout << arr[i] << " ";
	cout << endl;
}

int main(){
	int n;
	cout << "Enter array size :";
	cin >> n;
	int *arr = (int *)malloc(n * sizeof(int));
	for (int i = 0; i < n; i++){
		cout << "Enter Element : ";
		cin >> arr[i];
	}
	cout << endl << "Before Sort :" << endl;
	print(arr, n);
	merge_sort(arr, 0, n-1);
	cout << "After sort : " << endl;
	print(arr, n);
	return 0;
}
